native/src/main/java/io/ballerina/lib/data/xmldata/xml/XmlTraversal.java

```diff
@@ -83,29 +84,61 @@ static class XmlTree {
+            XmlAnalyzerData analyzerData = new XmlAnalyzerData();
+            DataUtils.updateOptions(options, analyzerData);
+            return traverseXml(xml, analyzerData, type);
+        }
+
+        public Object traverseXml(BXml xml, XmlAnalyzerData analyzerData, Type type) {
             Type referredType = TypeUtils.getReferredType(type);
             switch (referredType.getTag()) {
                 case TypeTags.RECORD_TYPE_TAG -> {
-                    XmlAnalyzerData analyzerData = new XmlAnalyzerData();
-                    DataUtils.updateOptions(options, analyzerData);
-                    RecordType recordType = (RecordType) referredType;
-                    currentNode = ValueCreator.createRecordValue(recordType.getPackage(), recordType.getName());
-                    BXml nextXml = validateRootElement(xml, recordType, analyzerData);
-                    Object resultRecordValue = traverseXml(nextXml, recordType, analyzerData);
-                    DataUtils.validateRequiredFields(analyzerData);
-                    return resultRecordValue;
+                    return traverseXmlWithRecordAsExpectedType(xml, analyzerData, referredType);
                 }
                 case TypeTags.MAP_TAG -> {
-                    MapType mapType = (MapType) referredType;
-                    RecordType anonRecType = TypeCreator.createRecordType(Constants.ANON_TYPE, mapType.getPackage(), 0,
-                            new HashMap<>(), mapType.getConstrainedType(), false, 0);
-                    return traverseXml(xml, options, anonRecType);
+                    return traverseXmlWithMapAsExpectedType(xml, referredType, analyzerData);
                 }
-                default -> {
-                    return DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, Constants.RECORD_OR_MAP,
-                            type.getName());
+                case TypeTags.UNION_TAG -> {
+                    return traverseXmlToUnion(xml, analyzerData, referredType);
+                }
+                default -> throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, Constants.RECORD_OR_MAP, type);
+            }
+        }
+        
+        private Object traverseXmlWithRecordAsExpectedType(BXml xml,
+                                                           XmlAnalyzerData analyzerData, Type referredType) {
+            RecordType recordType = (RecordType) referredType;
+            currentNode = ValueCreator.createRecordValue(recordType.getPackage(), recordType.getName());
+            BXml nextXml = validateRootElement(xml, recordType, analyzerData);
+            Object resultRecordValue = traverseXml(nextXml, recordType, analyzerData);
+            DataUtils.validateRequiredFields(analyzerData, (BMap<BString, Object>) currentNode);
+            return resultRecordValue;
+        }
+
+        private Object traverseXmlWithMapAsExpectedType(BXml xml, Type referredType, XmlAnalyzerData analyzerData) {
```

Comment:
Shall we make the parameter of concrete type
```suggestion
        private Object traverseXmlWithMapAsExpectedType(BXml xml, MapType referredType, XmlAnalyzerData analyzerData) {
```